README.md
# **rn-loading-wireframe**

It can help showing loading wireframe effect for its wrapped components.

## Features

- Support in React Native for both Android and iOS
- Possible to wrap more than one child component
- Easy to use
- Some customizations
- Produces wireframe with the size style corresponding the style of the wrapped child elements

## Demo

<img src=".github/demo.gif" height="500" />

## Installation

- Using **Npm**

```bash
npm install rn-loading-wireframe --save
```

- Using **Yarn**

```bash
yarn add rn-loading-wireframe
```

## Example

```jsx
import LoadingWireframe from 'rn-loading-wireframe';

<LoadingWireframe loading={true} color='#bdc3c7' fadeDuration={1000}>
  <Text
    style={{
      fontSize: 20,
      fontWeight: 'bold'
    }}>
    Loading Wireframe
  </Text>
  <Text
    style={{
      marginTop: 5
    }}>
    It can help showing loading wireframe effect
  </Text>
  <TouchableHighlight
    style={{
      marginTop: 10,
      width: 100,
      height: 40,
      justifyContent: 'center',
      alignItems: 'center',
      backgroundColor: 'orange',
      borderRadius: 6
    }}>
    <Text style={{ color: '#fff' }}>Download</Text>
  </TouchableHighlight>
</LoadingWireframe>;
```

## Usage

### Available props

| Name              | Type   | Default   | Description                        |
| ----------------- | ------ | --------- | ---------------------------------- |
| loading           | bool   | false     | Should show the loading wireframe? |
| color             | string | `#bdc3c7` | Color of the loading wireframe     |
| fadeDuration      | number | 1000      | Duration of fading blink           |
| lowestFadeOpacity | number | 0.2       | The lowest opacity when fading out |

## License

- [MIT](LICENSE)


CHANGELOG.md
# 1.0.2 (2019-09-02)

- Generate wireframes with correct size matching the wrapped children component event they are not specified size in style

# 1.0.0 (2019-08-27)

Initial version


src/LoadingWireframe.js
import React from 'react';
import {
  StyleSheet,
  ViewStyle,
  View,
  Animated,
  Easing,
  Dimensions
} from 'react-native';
import PropTypes from 'prop-types';

const defaultStyles = {
  wireframe: {
    backgroundColor: 'transparent',
    justifyContent: undefined,
    alignItems: undefined,
    overflow: 'hidden',
    borderWidth: undefined,
    borderTopWidth: undefined,
    borderBottomWidth: undefined,
    borderLeftWidth: undefined,
    borderRightWidth: undefined,
    borderStartWidth: undefined,
    borderEndWidth: undefined
  }
};

class LoadingWireframe extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      timingValue: new Animated.Value(0.0),
      targets: {}
    };

    const dimensions = Dimensions.get('screen');
    this.state.isPortrait = dimensions.width <= dimensions.height;
  }

  componentDidMount() {
    if (this.props.loading) {
      this._startFade();
    }
    Dimensions.addEventListener('change', this._onDimensionChanged);
  }

  componentWillUnmount() {
    Dimensions.removeEventListener('change', this._onDimensionChanged);
  }

  _onDimensionChanged = ({ screen }) => {
    if (!this.props.loading) {
      return;
    }

    this.setState(oldState => {
      const isPortrait = screen.width <= screen.height;

      if (isPortrait == oldState.isPortrait) {
        return null;
      }

      return {
        targets: {},
        isPortrait
      };
    });
  };

  componentDidUpdate(oldProps, oldState) {
    if (this.props.loading != oldProps.loading) {
      if (this.props.loading) {
        this._startFade();
      } else {
        this._stopFade();
      }
    }

    if (this.state.isPortrait != oldState.isPortrait) {
      // Animate fade again as after the dimension changed, it caused the previous animation to stop
      if (this.props.loading) {
        this._startFade();
      }
    }
  }

  _startFade() {
    this._stopFade();
    this.state.timingValue.setValue(0.0);
    this._anim = Animated.loop(
      Animated.timing(this.state.timingValue, {
        toValue: 1.0,
        duration: this.props.fadeDuration,
        easing: Easing.linear,
        useNativeDriver: true
      })
    );
    this._anim.start();
  }

  _stopFade() {
    this._anim && this._anim.stop();
    this._anim = null;
  }

  render() {
    const { loading, color, lowestFadeOpacity, children } = this.props;
    const opacity = this.state.timingValue.interpolate({
      inputRange: [0.0, 0.5, 1.0],
      outputRange: [1.0, lowestFadeOpacity, 1.0]
    });

    return React.Children.map(children, (child, index) => {
      return (
        <Wireframe
          key={index}
          opacity={opacity}
          color={color}
          loading={loading}
          size={this.state.targets[index] || null}
          onLayout={e => {
            this.setState(oldState => {
              return {
                targets: {
                  ...oldState.targets,
                  [index]: { width: e.layout.width, height: e.layout.height }
                }
              };
            });
          }}>
          {child}
        </Wireframe>
      );
    });
  }
}

const Wireframe = ({
  opacity,
  color,
  loading,
  children,
  size,
  onLayout,
  ...restProps
}) => {
  if (!loading || !React.isValidElement(children)) {
    return children;
  }

  let wireframeStyle: ViewStyle = !children.props.style
    ? defaultStyles.wireframe
    : {
        ...StyleSheet.flatten(children.props.style),
        ...defaultStyles.wireframe
      };

  const hasProvidedFixedSize =
    wireframeStyle.width !== null &&
    wireframeStyle.width !== undefined &&
    wireframeStyle.height !== null &&
    wireframeStyle.height !== undefined;

  if (size || hasProvidedFixedSize) {
    if (!hasProvidedFixedSize) {
      wireframeStyle = {
        ...wireframeStyle,
        width: size.width,
        height: size.height
      };
    }

    return (
      <View {...restProps} style={wireframeStyle}>
        <Animated.View
          style={{
            backgroundColor: color,
            flex: 1,
            opacity: opacity
          }}
        />
      </View>
    );
  } else if (onLayout) {
    return React.cloneElement(children, {
      onLayout: e => {
        onLayout && onLayout(e.nativeEvent);
      },
      style: { ...wireframeStyle, opacity: 0 }
    });
  }
};

LoadingWireframe.defaultProps = {
  lowestFadeOpacity: 0.2,
  fadeDuration: 1000,
  color: '#bdc3c7',
  loading: false
};

LoadingWireframe.propTypes = {
  lowestFadeOpacity: PropTypes.number.isRequired,
  fadeDuration: PropTypes.number.isRequired,
  color: PropTypes.string.isRequired,
  loading: PropTypes.bool.isRequired
};

export default LoadingWireframe;


examples/App.js
/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * @format
 * @flow
 * @lint-ignore-every XPLATJSCOPYRIGHT1
 */

import React from 'react';
import {
  SafeAreaView,
  View,
  StyleSheet,
  Text,
  Image,
  TouchableHighlight,
  ScrollView
} from 'react-native';
import LoadingWireframe from '../src';

const loadingWireframeColor = '#bdc3c7';
const loadingWireframeFadeDuration = 1000;

export default class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      loading: true
    };
  }

  componentDidMount() {
    setTimeout(() => {
      this.setState({ loading: false });
    }, 2000);
  }

  render() {
    const { loading } = this.state;

    return (
      <SafeAreaView style={styles.container}>
        <ScrollView
          contentContainerStyle={{ paddingTop: 10, paddingHorizontal: 10 }}>
          {[1, 2, 3, 4, 5, 6].map((v, i) => {
            return (
              <View key={i} style={styles.card}>
                <LoadingWireframe
                  loading={loading}
                  color={loadingWireframeColor}
                  fadeDuration={loadingWireframeFadeDuration}>
                  <Image
                    style={styles.image}
                    source={{
                      uri:
                        'https://facebook.github.io/react-native/docs/assets/favicon.png'
                    }}
                  />
                </LoadingWireframe>
                <View style={{ flex: 1 }}>
                  <LoadingWireframe
                    loading={loading}
                    color={loadingWireframeColor}
                    fadeDuration={loadingWireframeFadeDuration}>
                    <View
                      style={{
                        borderBottomWidth: 1,
                        borderBottomColor: 'red'
                      }}>
                      <Text
                        style={{
                          fontSize: 20,
                          fontWeight: 'bold'
                        }}>
                        Loading Wireframe
                      </Text>
                    </View>

                    <View
                      style={{
                        marginTop: 5
                      }}>
                      <Text>It can help showing loading wireframe effect</Text>
                    </View>

                    <TouchableHighlight style={styles.button}>
                      <Text style={{ color: '#fff' }}>Download</Text>
                    </TouchableHighlight>
                  </LoadingWireframe>
                </View>
              </View>
            );
          })}
        </ScrollView>
      </SafeAreaView>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff'
  },
  card: {
    flexDirection: 'row',
    borderRadius: 6,
    marginBottom: 10,
    padding: 10,
    backgroundColor: '#fff',
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 1
    },
    shadowOpacity: 0.22,
    shadowRadius: 2.22,
    elevation: 3
  },
  image: {
    width: 100,
    height: 100,
    marginBottom: 10,
    borderRadius: 50,
    marginEnd: 10
  },
  button: {
    position: 'absolute',
    end: 0,
    bottom: 0,
    width: 100,
    height: 40,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'orange',
    borderRadius: 6
  }
});
